File: src/main/java/com/joseestudillo/spark/sql/DataFrameSparkSQL.java
package com.joseestudillo.spark.sql;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.apache.log4j.Logger;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import com.joseestudillo.spark.sql.DataFrameSparkSQL.DataBean;
import com.joseestudillo.spark.utils.SparkUtils;

/**
 * Example of loading data into DataFrames from RDDs
 * 
 * @author Jose Estudillo
 *
 */
public class DataFrameSparkSQL {

	private static final Logger log = Logger.getLogger(DataFrameSparkSQL.class);

	private static final String TABLE_NAME = "databean_table";

	//bean to hold the data read from the input RDD. this class must be Serializable.
	public static class DataBean implements Serializable {

		private static final long serialVersionUID = 1L;

		private String name;
		private long id;

		public DataBean(String name, long id) {
			this.name = name;
			this.id = id;
		}

		public String getName() {
			return name;
		}

		public void setName(String name) {
			this.name = name;
		}

		public long getId() {
			return id;
		}

		public void setId(long id) {
			this.id = id;
		}

		@Override
		public String toString() {
			return "DataBean [name=" + name + ", id=" + id + "]";
		}
	}

	public static void main(String[] args) {
		SparkConf conf = SparkUtils.getLocalConfig(DataFrameSparkSQL.class.getSimpleName());
		log.info(String.format("access to the web interface at localhost: %s", SparkUtils.SPARK_UI_PORT));
		JavaSparkContext sparkContext = new JavaSparkContext(conf);

		SQLContext sqlContext = new SQLContext(sparkContext);

		List<String> inputList = Arrays.asList("abc, 1", "def, 2", "ghi, 3");
		JavaRDD<String> inputRdd = sparkContext.parallelize(inputList);

		// # Read data into a data frame using beans
		org.apache.spark.api.java.function.Function<String, DataBean> fromTextToDataBean = s -> {
			String[] tokens = s.split(",");
			return new DataBean(tokens[0], Long.parseLong(tokens[1].trim()));
		};

		// transform the plain text into DataBeans
		JavaRDD<DataBean> dataBeanRDD = inputRdd.map(fromTextToDataBean);
		log.info(String.format("Transforming the input %s into dataBeans %s", inputRdd.collect(), dataBeanRDD.collect()));

		// Create a DataFrame from the DataBeans RDD
		DataFrame dataBeanDataFrame = sqlContext.createDataFrame(dataBeanRDD, DataBean.class);
		dataBeanDataFrame.registerTempTable(TABLE_NAME);
		log.info(String.format("Table '%s' generated from %s:", TABLE_NAME, dataBeanRDD.collect()));
		dataBeanDataFrame.show();

		// Get all the rows in the dataFrame
		JavaRDD<Row> rddFromDataFrame = dataBeanDataFrame.javaRDD();

		// log the rows in a JSON format
		Function<Row, String> fromRowToString = row -> String.format("Row: {id:%s, name:%s}", row.getLong(0), row.getString(1));
		log.info(String.format("Transforming rowRdd %s into strings: %s", rddFromDataFrame.collect(), rddFromDataFrame.map(fromRowToString).collect()));

		// # Read data into a dataframe using plain data and defining the schema
		// define the column names/types
		List<StructField> columnTypes = new ArrayList<StructField>();
		columnTypes.add(DataTypes.createStructField("id", DataTypes.LongType, true));
		columnTypes.add(DataTypes.createStructField("name", DataTypes.StringType, true));
		StructType tableSchema = DataTypes.createStructType(columnTypes);

		log.info(String.format("Table Squema: %s", tableSchema));

		// Create rows (Row) from plain text
		Function<String, Row> fromStringToRow = inputText -> {
			String[] fields = inputText.split(",");
			return (Row) RowFactory.create(fields[1].trim(), fields[0]);
		};
		JavaRDD<Row> rowsFromDataRDD = inputRdd.map(fromStringToRow);
		log.info(String.format("inputRdd %s -> outputRowRdd %s", inputRdd.collect(), rowsFromDataRDD.collect()));

		// Apply the schema to the RDD and load the Rows from the RDD.
		DataFrame definedSchemaDataFrame = sqlContext.createDataFrame(rowsFromDataRDD, tableSchema);
		definedSchemaDataFrame.registerTempTable(TABLE_NAME);
		log.info(String.format("Table '%s' generated from %s:", TABLE_NAME, rowsFromDataRDD.collect()));
		definedSchemaDataFrame.show();

	}
}

File: src/main/scala/com/joseestudillo/spark_scala/sql/DataFrameSparkSQL.scala
package com.joseestudillo.spark_scala.sql

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import com.joseestudillo.spark_scala.utils.SparkUtils
import org.slf4j.LoggerFactory
import com.typesafe.scalalogging.Logger
import org.apache.spark.sql.SQLContext
import org.apache.spark.rdd.RDD
import org.apache.spark.api.java.JavaRDD
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.types.DataTypes
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.types.StructField
import org.apache.spark.sql.Row

object DataFrameSparkSQL {
  
  val log = Logger(LoggerFactory.getLogger(getClass.getName))
  
  val TABLE_NAME = "databean_table"
  
  def main(args: Array[String]) {
    
    class DataBean(var name: String,var id: Long) extends Serializable {
      override def toString = s"DataBean($name, $id)"
      def setName(s: String) { name = s }
      def setId(l: Long) { id = l }
      def getName = name
      def getId = id
    }
    
    val host = if (args.length > 0) args(0) else SparkUtils.LOCAL_MASTER_ID
    val appName = getClass.getName

    val conf = new SparkConf().setAppName(appName).setMaster(host)
    val sparkContext = new SparkContext(conf)
    val sparkSQLContext = new SQLContext(sparkContext)


    val inputList = List("abc, 1", "def, 2", "ghi, 3")
		val inputRdd: RDD[String] = sparkContext.parallelize(inputList)

		val toDataBean: String => DataBean = x => {
		  val tokens = x.split(", ") 
		  new DataBean(tokens(0), tokens(1).toLong)
		}
		
		// transform the plain text into DataBeans
		val dataBeanRDD: RDD[DataBean] = inputRdd.map(toDataBean)
		log.info(s"Transforming the input ${inputRdd.collect().mkString(", ")} into dataBeans ${dataBeanRDD.collect().mkString(", ")}")

		// Create a DataFrame from the DataBeans RDD
		val dataBeanDataFrame = sparkSQLContext.createDataFrame(dataBeanRDD, classOf[DataBean])
		dataBeanDataFrame.registerTempTable(TABLE_NAME)
		log.info(s"Table '$TABLE_NAME' generated from ${dataBeanRDD.collect().mkString(", ")}:")
		dataBeanDataFrame.show()

		// Get all the rows in the dataFrame
		val rddFromDataFrame: RDD[Row] = dataBeanDataFrame.rdd

		// log the rows in a JSON format
		val fromRowToString = (row: Row) => {
		  s"Row: {id:${row.getLong(0)}, name:${row.getString(1)}}"
		}
		log.info(s"Transforming rowRdd ${rddFromDataFrame.collect().mkString(", ")} into strings: ${rddFromDataFrame.map(fromRowToString).collect().mkString(", ")}")

		// # Read data into a dataframe using plain data and defining the schema
		// define the column names/types
		val columnTypes: Array[StructField] = new Array[StructField](2)
		columnTypes.update(0, DataTypes.createStructField("name", DataTypes.StringType, true))
		columnTypes.update(1, DataTypes.createStructField("id", DataTypes.LongType, true))
		val tableSchema: StructType = DataTypes.createStructType(columnTypes)

		log.info(s"Table Squema: $tableSchema")

		// Create rows (Row) from plain text
		val fromStringToRow = (inputText: String) => {
			val fields = inputText.split(",")
			Row(fields(0).trim(), fields(1).trim().toLong)
		}
		val rowsFromDataRDD: RDD[Row] = inputRdd.map(fromStringToRow)
		log.info(s"inputRdd ${inputRdd.collect().mkString(", ")} -> outputRowRdd ${rowsFromDataRDD.collect().mkString(", ")}")

		// Apply the schema to the RDD and load the Rows from the RDD.
		val definedSchemaDataFrame: DataFrame = sparkSQLContext.createDataFrame(rowsFromDataRDD, tableSchema)
		definedSchemaDataFrame.registerTempTable(TABLE_NAME)
		log.info(s"Table '$TABLE_NAME' generated from ${rowsFromDataRDD.collect().mkString(", ")}:")
		definedSchemaDataFrame.show()
    
    sparkContext.stop()
  }
}